components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/device/grant/DeviceFlowGrant.java

```diff
@@ -0,0 +1,167 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.oauth2.device.grant;
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.identity.application.authentication.framework.model.AuthenticatedUser;
+import org.wso2.carbon.identity.oauth2.IdentityOAuth2Exception;
+import org.wso2.carbon.identity.oauth2.device.constants.Constants;
+import org.wso2.carbon.identity.oauth2.device.dao.DeviceFlowPersistenceFactory;
+import org.wso2.carbon.identity.oauth2.device.errorcodes.DeviceErrorCodes;
+import org.wso2.carbon.identity.oauth2.device.model.DeviceFlowDO;
+import org.wso2.carbon.identity.oauth2.dto.OAuth2AccessTokenRespDTO;
+import org.wso2.carbon.identity.oauth2.model.RequestParameter;
+import org.wso2.carbon.identity.oauth2.token.OAuthTokenReqMessageContext;
+import org.wso2.carbon.identity.oauth2.token.handlers.grant.AbstractAuthorizationGrantHandler;
+import org.wso2.carbon.identity.oauth2.util.OAuth2Util;
+
+import java.sql.Timestamp;
+import java.util.Date;
+
+/**
+ * Device flow grant type for Identity Server.
+ */
+public class DeviceFlowGrant extends AbstractAuthorizationGrantHandler {
+
+    private static Log log = LogFactory.getLog(DeviceFlowGrant.class);
+
+    @Override
+    public boolean validateGrant(OAuthTokenReqMessageContext oAuthTokenReqMessageContext) throws
+            IdentityOAuth2Exception {
+
+        super.validateGrant(oAuthTokenReqMessageContext);
+        boolean authStatus = false;
+        RequestParameter[] parameters = oAuthTokenReqMessageContext.getOauth2AccessTokenReqDTO()
+                .getRequestParameters();
+        String deviceCode = null;
+        String deviceStatus;
+
+        for (RequestParameter parameter : parameters) {
+            if (Constants.DEVICE_CODE.equals(parameter.getKey()) && parameter.getValue() != null
+                    && parameter.getValue().length > 0) {
+                deviceCode = parameter.getValue()[0];
+                break;
+            }
+        }
+
+        if (log.isDebugEnabled()) {
+            log.debug("Getting ready to release token for device_code: " + deviceCode);
+        }
+
+        if (deviceCode != null) {
+            DeviceFlowDO deviceFlowDO = DeviceFlowPersistenceFactory.getInstance().getDeviceFlowDAO()
+                    .getAuthenticationDetails(deviceCode);
+            Date date = new Date();
+            deviceStatus = deviceFlowDO.getStatus();
+            deviceFlowDO.setScope(getScopes(deviceCode));
```

Comment:
Can we reduce this DB call by getting the scope details also from `getAuthenticationDetails` ?

Reply:
Those are in two different tables. If we want to do it, we need to join the table before get values.